Tidy check for test revisions that are mentioned but not declared

Zalathar opened the pull request:
If a `[revision]` name appears in a test header directive or error annotation, but isn't declared in the `//@ revisions:` header, that is almost always a mistake.

In cases where a revision needs to be temporarily disabled, adding it to an `//@ unused-revision-names:` header will suppress these checks for that name.

Adding the wildcard name `*` to the unused list will suppress these checks for the entire file.

(None of the tests actually use `*`; it's just there because it was easy to add and could be handy as an escape hatch when dealing with other problems.)

---

Most of the existing problems discovered by this check were fairly straightforward to fix (or ignore); the trickiest cases are in `borrowck` tests.

Zalathar committed:
Tidy check for test revisions that are mentioned but not declared

If a `[revision]` name appears in a test header directive or error annotation,
but isn't declared in the `//@ revisions:` header, that is almost always a
mistake.

In cases where a revision needs to be temporarily disabled, adding it to an
`//@ unused-revision-names:` header will suppress these checks for that name.

Adding the wildcard name `*` to the unused list will suppress these checks for
the entire file.

Zalathar committed:
Include the line number in tidy's `iter_header`

Zalathar committed:
Fix test problems discovered by the revision check

Most of these changes either add revision names that were apparently missing,
or explicitly mark a revision name as currently unused.

Zalathar commented on src/tools/compiletest/src/header.rs:
```diff
@@ -925,6 +925,7 @@ const KNOWN_DIRECTIVE_NAMES: &[&str] = &[
     "test-mir-pass",
     "unset-exec-env",
     "unset-rustc-env",
+    ("unused-revision-names"/* used by tidy check `unknown_revision` */),
```
These parentheses are here because if I just add a normal end-of-line comment, `rustfmt` insists on reindenting the `// tidy-alphabetical-end` to be aligned with it.

I'm open to other suggestions.

Urgau replied:
Wouldn't this work?
```suggestion
    // used by tidy check `unknown_revision`
    "unused-revision-names",
```

Zalathar replied:
Oh, you might be right. I had mentally ruled it out for reasons that aren’t relevant in this context.

jieyouxu replied:
Also what I have in mind for compiletest directive parsing is that if we rework the parsing to stop silently allowing unknown directives to go through, we wouldn't need this hacky `KNOWN_DIRECTIVE_NAMES` array which isn't exhaustive anyway and requires people to modify it every time a directive is added or removed or renamed.

Zalathar commented:
It would be nice to have a variant of this for Filecheck directives as well, but that's probably trickier because they're more freeform and thus harder to identify reliably.

bors commented:
:umbrella: The latest upstream changes (presumably #124345) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

jieyouxu reviewed (commented):
I feel like the long-term solution to revision-related funny business is to rework how compiletest directive collection, parsing and validation is done. This would include adding validation for revisions that are used but not declared (i.e. the problem this PR is trying to solve), duplicate, invalid (incl. invalid characters and such), etc.

I think having a tidy check in the meantime (before compiletest directive rework, which I plan to start work on towards the end of this month) is reasonable though. It just feels like it should be the job of compiletest rather than a tidy check.

jieyouxu commented:
I'll roll a bootstrap reviewer as well because this mostly concerns tidy which falls under the purview of bootstrap. I think this PR is definitely addressing a "wtf moment" when writing and reading tests, I'm just not sure if it should be tidy's responsibility.

r? bootstrap

Zalathar commented:
Yeah, my original implementation was integrated into compiletest, but it felt like I was adding more mess that would make future cleanups even trickier. So I decided to try a more self-contained tidy-based version instead.

So I’m fine with this being seen as a relatively temporary solution, to eventually be replaced with better checks inside compiletest itself.

jieyouxu commented:
Also see more silliness for compiletest directive/revision handling:
- https://github.com/rust-lang/rust/issues/123765
- https://github.com/rust-lang/rust/issues/123760

Mark-Simulacrum commented:
r? jieyouxu -- tidy seems like a reasonable place

jieyouxu commented on src/tools/tidy/src/unknown_revision.rs:
Should we leave a FIXME here (feel free to `FIXME(jieyouxu)`) to remind people (well, mostly me) that ideally we would have these checks as part of compiletest directive and revision handling?

jieyouxu reviewed (approved):
The implementation for this looks good to me. One minor nit about maybe leaving a FIXME for the tidy check to not forget to replace it with proper handling in compiletest in the long term.

jieyouxu commented:
Feel free to r=me with or without the FIXME.

@bors delegate=Zalathar

bors commented:
:v: @Zalathar, you can now approve this pull request!

If @jieyouxu told you to "`r=me`" after making some further change, please make that change, then do `@bors r=@jieyouxu`
<!-- homu: {"type":"Delegated","delegator":"jieyouxu","delegate":"Zalathar"} -->

Zalathar commented:
Rebased and added FIXME [(diff)](https://github.com/rust-lang/rust/compare/80b67c85eefeaaab38c798b57a639a114db96bad..14d56e833848b94c8c257c53886d0331a4ac735e).

@bors r=jieyouxu

bors commented:
:pushpin: Commit 14d56e833848b94c8c257c53886d0331a4ac735e has been approved by `jieyouxu`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=jieyouxu 14d56e833848b94c8c257c53886d0331a4ac735e -->
<!-- homu: {"type":"Approved","sha":"14d56e833848b94c8c257c53886d0331a4ac735e","approver":"jieyouxu","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit 14d56e833848b94c8c257c53886d0331a4ac735e with merge 8c7c151a7a03d92cc5c75c49aa82a658ec1fe4ff...
<!-- homu: {"type":"BuildStarted","head_sha":"14d56e833848b94c8c257c53886d0331a4ac735e","merge_sha":"8c7c151a7a03d92cc5c75c49aa82a658ec1fe4ff"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/9017757327/job/24782204976)
Approved by: jieyouxu
Pushing 8c7c151a7a03d92cc5c75c49aa82a658ec1fe4ff to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"jieyouxu","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/9017757327/job/24782204976"},"merge_sha":"8c7c151a7a03d92cc5c75c49aa82a658ec1fe4ff"} -->

rust-timer commented:
Finished benchmarking commit (8c7c151a7a03d92cc5c75c49aa82a658ec1fe4ff): [comparison URL](https://perf.rust-lang.org/compare.html?start=6f7e00a3e6eade8350ce6280f05596236d73c4cb&end=8c7c151a7a03d92cc5c75c49aa82a658ec1fe4ff&stat=instructions:u).

### Overall result: ❌✅ regressions and improvements - no action needed


@rustbot label: -perf-regression



### [Instruction count](https://perf.rust-lang.org/compare.html?start=6f7e00a3e6eade8350ce6280f05596236d73c4cb&end=8c7c151a7a03d92cc5c75c49aa82a658ec1fe4ff&stat=instructions:u)
This is a highly reliable metric that was used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | 0.4%  | [0.3%, 0.5%]   | 5     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -0.4% | [-0.4%, -0.4%] | 1     |
| All ❌✅ (primary)                 | -     | -              | 0     |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=6f7e00a3e6eade8350ce6280f05596236d73c4cb&end=8c7c151a7a03d92cc5c75c49aa82a658ec1fe4ff&stat=max-rss)
This benchmark run did not return any relevant results for this metric.

### [Cycles](https://perf.rust-lang.org/compare.html?start=6f7e00a3e6eade8350ce6280f05596236d73c4cb&end=8c7c151a7a03d92cc5c75c49aa82a658ec1fe4ff&stat=cycles:u)
This benchmark run did not return any relevant results for this metric.

### [Binary size](https://perf.rust-lang.org/compare.html?start=6f7e00a3e6eade8350ce6280f05596236d73c4cb&end=8c7c151a7a03d92cc5c75c49aa82a658ec1fe4ff&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 673.446s -> 674.785s (0.20%)
**Artifact size**: 315.79 MiB -> 315.96 MiB (0.05%)
